### PR TITLE
Adds lifetime to IoCmd, init ipmi commands

### PR DESCRIPTION
**Pull request by SuchAFuriousDeath**

### CODE REVIEW

**Comment by SuchAFuriousDeath** [comment]
@ultimaweapon Does it make sense for a each process to have it's own instance of Vm? shouldn't it be inside of the ProcManager?

**Comment by ultimaweapon** [comment]
That is not possible on macOS. It also seems like not a good idea since the amount of process may be hundred.

**Comment by SuchAFuriousDeath** [comment]
If it doesn't make sense... why is it that way in the code? :eyes:

**Comment by SuchAFuriousDeath** [comment]
Well... I suppose it made sense with only one process

File: src/kernel/src/fs/ioctl.rs [hunk 1]
```diff
@@ -91,7 +91,7 @@ macro_rules! commands {
 // TODO: implement void ioctl commands with int data
 
 commands! {
-    pub enum IoCmd {
+    pub enum IoCmd<'a> {
```
Comment (ultimaweapon) [inline src/kernel/src/fs/ioctl.rs:94]: This lifetime need to be referenced on each variant too.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

**Comment by ultimaweapon** [comment]
> If it doesn't make sense... why is it that way in the code? 👀

What do you mean by that? Our code current have a single VM that will be hosting multiple processes.

**Comment by SuchAFuriousDeath** [comment]
I mean the Vm struct

**Comment by ultimaweapon** [comment]
In that case yes, each process will have only one. I believed the reason FreeBSD put this as a top-level is because the VM can live alone without a process. The process is just one of a thing that depend on the VM. So I think it is better to keep it on a top-level.

File: src/kernel/src/process/mod.rs [hunk 2]
```diff
@@ -382,16 +382,19 @@ impl ProcManager {
...
     fn sys_cpuset_setaffinity(self: &Arc<Self>, _: &VThread, i: &SysIn) -> Result<SysOut, SysErr> {
+        let level: CpuLevel = i.args[0].try_into()?;
+        let which: CpuWhich = i.args[1].try_into()?;
+        let id: i64 = i.args[2].into();
         let cpusetsize: usize = i.args[3].into();
+        let mask: *const u8 = i.args[4].into();
 
-        // TODO: Refactor this for readability.
-        if cpusetsize.wrapping_sub(8) > 8 {
+        if cpusetsize.wrapping_sub(8) > std::mem::size_of::<CpuSet>() {
```
Comment (ultimaweapon) [inline src/kernel/src/process/mod.rs:391]: Our `CpuSet` is not `cpuset_t` so this is not correct.

File: src/kernel/src/process/mod.rs [hunk 3]
```diff
@@ -346,7 +346,7 @@ impl ProcManager {
         let mask: *mut u8 = i.args[4].into();
 
         // TODO: Refactor this for readability.
-        if cpusetsize.wrapping_sub(8) > 8 {
+        if cpusetsize.wrapping_sub(8) > std::mem::size_of::<CpuSet>() {
```
Comment (ultimaweapon) [inline src/kernel/src/process/mod.rs:349]: Same here.

File: src/kernel/src/ipmi/mod.rs [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+impl IpmiManager {
+    pub fn new(syscalls: &mut Syscalls) -> Arc<Self> {
+        let ipmi = Arc::new(Self {});
+
+        syscalls.register(622, &ipmi, Self::ipmi_mgr_call);
+
+        ipmi
+    }
+
+    fn ipmi_mgr_call(self: &Arc<Self>, td: &VThread, i: &SysIn) -> Result<SysOut, SysErr> {
```
Comment (ultimaweapon) [inline src/kernel/src/ipmi/mod.rs:24]: All syscall handlers should have the same naming convention.

**Comment by SuchAFuriousDeath** [comment]
I'm not sure I can make us run in a VM. Maybe you have to do it

File: src/kernel/src/vm/mod.rs [hunk 5]
```diff
@@ -622,7 +622,7 @@ impl Vm {
             return Err(SysErr::Raw(EINVAL));
         }
 
-        let slice_size = num_of_ops.try_into().ok().ok_or(SysErr::Raw(EINVAL))?;
+        let slice_size = num_of_ops.try_into().map_err(|_| SysErr::Raw(EINVAL))?;
```
Comment (ultimaweapon) [inline src/kernel/src/vm/mod.rs:625]: I can't find this check on the PS4.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: src/kernel/src/vm/mod.rs (hunk 5 above)
Comment (SuchAFuriousDeath) [reply src/kernel/src/vm/mod.rs]: image

I refactored it

**Review by ultimaweapon — APPROVED** [review]

**Pull request merged**